boa/src/builtins/number/mod.rs

```diff
@@ -277,25 +318,99 @@ impl Number {
-                x if x > 0 => n.to_integer(context)? as usize,
-                _ => 0,
-            },
-            None => 0,
+        let precision_var = args.get(0).cloned().unwrap_or_default();
+
+        // 1 & 6
+        let mut this_num = Self::this_number_value(this, context)?;
+        // 2 & 4
+        if precision_var == Value::undefined() || !this_num.is_finite() {
+            return Self::to_string(this, &[], context);
+        }
+
+        // 3
+        let precision = match precision_var.to_integer_or_infinity(context).unwrap() {
+            IntegerOrInfinity::Integer(x) if (1..=100).contains(&x) => x as usize,
+            _ => {
+                // 5
+                return context.throw_range_error(
+                    "precision must be an integer at least 1 and no greater than 100",
+                );
+            }
         };
-        // TODO: Implement toPrecision
-        unimplemented!("TODO: Implement toPrecision");
+        let precision_i32 = precision as i32;
+
+        // 7
+        let mut prefix = String::new(); // spec: "s"
+        let mut suffix: String; // spec: "m"
+        let exponent: i32; // spec: "e"
+
+        // 8
+        if this_num < 0.0 {
+            prefix.push('-');
+            this_num = -this_num;
+        }
+
+        let mut is_scientific = false;
+
+        // 9
+        if this_num == 0.0 {
+            suffix = "0".repeat(precision);
+            exponent = 0;
+        // 10
+        } else {
+            let mut buffer = ryu_js::Buffer::new();
+            suffix = buffer.format(this_num).to_string();
+
+            exponent = Self::flt_str_to_exp(&suffix);
+            if exponent < 0 {
+                suffix = suffix.split_off((1 - exponent) as usize);
+            } else if let Some(n) = suffix.find('.') {
+                suffix.remove(n);
+            }
+            Self::round_to_precision(&mut suffix, precision);
+
+            let great_exp = exponent >= precision_i32;
+            if exponent < -6 || great_exp {
+                is_scientific = true;
+                suffix.truncate(precision);
+                if precision > 1 {
+                    suffix.insert(1, '.');
+                }
+                suffix.push('e');
+                if great_exp {
+                    suffix.push('+');
+                }
+                suffix.push_str(&exponent.to_string());
+            }
+        }
+
+        // 11
+        let e_inc = exponent + 1;
+        if !is_scientific && e_inc != precision_i32 {
+            // 12
+            if exponent >= 0 {
+                suffix.truncate(precision);
+                if precision > e_inc as usize {
+                    suffix.insert(e_inc as usize, '.');
+                }
+            // 13
+            } else {
+                prefix.push('0');
+                prefix.push('.');
+                prefix.push_str(&"0".repeat(-e_inc as usize));
+            }
+        }
```

Comment:
As a last thing I'd preffer if you followed the control flow from the spec, 11 is a separate `if` in the spec with an early return.